rust/crates/sdk/src/stream.rs

```diff
@@ -0,0 +1,286 @@
+    feed_ids: Vec<ID>,
+    conn: Option<WebSocketConnection>,
+    report_sender: mpsc::Sender<WebSocketReport>,
+    report_receiver: mpsc::Receiver<WebSocketReport>,
+    shutdown_sender: broadcast::Sender<()>,
+    stats: Arc<Stats>,
+    water_mark: Arc<Mutex<HashMap<String, usize>>>,
+}
+
+impl Stream {
+    /// Establishes a streaming WebSocket connection that sends reports for the given feedID(s) after they are verified.
+    ///
+    /// # Arguments
+    ///
+    /// * `config` - A validated `Config` instance.
+    /// * `feedIDs` - A comma-separated list of Data Streams feed IDs.
+    ///
+    /// # Endpoint:
+    /// ```bash
+    /// /api/v1/ws
+    /// ```
+    ///
+    /// # Type:
+    /// * WebSocket
+    ///
+    /// # Sample Request:
+    /// ```bash
+    /// GET /api/v1/ws?feedIDs=<feedID1>,<feedID2>,...
+    /// ```
+    ///
+    /// # Sample Response:
+    /// ```json
+    /// {
+    ///     "report": {
+    ///         "feedID": "Hex encoded feedId.",
+    ///         "fullReport": "A blob containing the report context and body. Encode the fee token into the payload before passing it to the contract for verification.",
+    ///         "validFromTimestamp": "Report's earliest applicable timestamp (in seconds).",
+    ///         "observationsTimestamp": "Report's latest applicable timestamp (in seconds)."
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// # Error Response Codes
+    ///
+    /// | Status Code | Description |
+    /// |-------------|-------------|
+    /// | **400 Bad Request** | This error is triggered when:<br>- There is any missing/malformed query argument.<br>- Required headers are missing or provided with incorrect values. |
+    /// | **401 Unauthorized User** | This error is triggered when:<br>- Authentication fails, typically because the HMAC signature provided by the client doesn't match the one expected by the server.<br>- A user requests access to a feed without the appropriate permission or that does not exist. |
+    /// | **500 Internal Server** | Indicates an unexpected condition encountered by the server, preventing it from fulfilling the request. This error typically points to issues on the server side. |
+    pub async fn new(config: &Config, feed_ids: Vec<ID>) -> Result<Stream, StreamError> {
+        let (report_sender, report_receiver) = mpsc::channel(100);
+        let (shutdown_sender, _) = broadcast::channel(1);
+
+        let stats = Arc::new(Stats {
+            accepted: AtomicUsize::new(0),
+            deduplicated: AtomicUsize::new(0),
+            partial_reconnects: AtomicUsize::new(0),
+            full_reconnects: AtomicUsize::new(0),
+            configured_connections: AtomicUsize::new(0),
+            active_connections: AtomicUsize::new(0),
+        });
+
+        let conn = connect(config, &feed_ids, stats.clone()).await?;
+
+        let water_mark = Arc::new(Mutex::new(HashMap::new()));
+
+        Ok(Stream {
+            config: config.clone(),
```

Comment:
Nit: A slightly better practice is to pass the config by value instead of by reference and cloning. Then the caller can decide whether to clone or to sink the config here if they have no further use for it.